crypto/src/signature.rs

```diff
@@ -0,0 +1,85 @@
+// SPDX-FileCopyrightText: 2023 TriliTech <contact@trili.tech>
+// Copyright (c) SimpleStaking, Viable Systems and Tezedge Contributors
+//
+// Ported from octez: lib_crypto/signature_v1.ml
+//
+// Copyright (c) 2018 Dynamic Ledger Solutions, Inc. <contact@tezos.com>
+// Copyright (c) 2020 Metastate AG <hello@metastate.dev>
+// Copyright (c) 2022 Nomadic Labs <contact@nomadic-labs.com>
+//
+// SPDX-License-Identifier: MIT
+
+use crate::base58::FromBase58CheckError;
+use crate::hash::{
+    BlsSignature, Ed25519Signature, FromBytesError, HashTrait, HashType, P256Signature,
+    Secp256k1Signature, UnknownSignature,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Signature {
+    Ed25519(Ed25519Signature),
+    Secp256k1(Secp256k1Signature),
+    P256(P256Signature),
+    Bls(BlsSignature),
+    Unknown(UnknownSignature),
+}
+
+impl Signature {
+    #[doc(hidden)]
+    pub fn from_vec(hash: Vec<u8>) -> Result<Self, FromBytesError> {
+        if hash.len() == HashType::BlsSignature.size() {
+            Ok(Signature::Bls(BlsSignature(hash)))
+        } else if hash.len() == HashType::UnknownSignature.size() {
+            Ok(Signature::Unknown(UnknownSignature(hash)))
+        } else {
+            Err(FromBytesError::InvalidSize)
+        }
+    }
+
+    pub fn from_base58_check(data: &str) -> Result<Self, FromBase58CheckError> {
+        if data.starts_with("edsig") {
+            Ok(Signature::Ed25519(Ed25519Signature::from_b58check(data)?))
+        } else if data.starts_with("spsig1") {
+            Ok(Signature::Secp256k1(Secp256k1Signature::from_b58check(
+                data,
+            )?))
+        } else if data.starts_with("p2sig") {
+            Ok(Signature::P256(P256Signature::from_b58check(data)?))
+        } else if data.starts_with("BLsig") {
+            Ok(Signature::Bls(BlsSignature::from_b58check(data)?))
+        } else {
+            Ok(Signature::Unknown(UnknownSignature::from_b58check(data)?))
+        }
+    }
+
+    pub fn to_base58_check(&self) -> String {
+        match self {
+            Self::Ed25519(s) => s.to_b58check(),
+            Self::Secp256k1(s) => s.to_b58check(),
+            Self::P256(s) => s.to_b58check(),
+            Self::Bls(s) => s.to_b58check(),
+            Self::Unknown(s) => s.to_b58check(),
+        }
+    }
+}
+
+impl AsRef<[u8]> for Signature {
```

Comment:
This `AsRef<[u8]>` kinda-sorta clashes with other "hash types", which implement `AsRef<Vec<u8>>` (aka `AsRef<Hash>`). Now, I think `AsRef<Vec<_>>` is generally a mistake: you can't do anything with a `&Vec` you couldn't do with a slice, and slice is more general. So by no means I suggest changing this instance to `AsRef<Hash>`.

But maybe it's a good excuse to change `AsRef<Hash>` to `AsRef<[u8]>` in `crypto/src/hash.rs`?

Reply:
done

Reply:
Nice, thanks.